### PR TITLE
[Scheduler] Align `RadixCache` no-insert cleanup with `kv_len_to_handle`

### PR DESCRIPTION
**Pull request by nanjiangwill**
## Summary

- Size the no-insert release from `kv_len_to_handle` instead of the committed token IDs, matching what every other cache backend already does.
- Skip the radix-key construction and value copy on the no-insert path; only the insertion path needs them.
- Related: #35793 bounds the cacheable length on the insertion path.

## Background

- `cache_finished_req(..., is_insert=False)` discards a request without inserting its KV. The old code built `token_ids = (origin_input_ids + output_ids)[:kv_len_to_handle]` and released `[cache_protected_len, len(token_ids))`.
- `swa_radix_cache`, `pure_swa_radix_cache`, `mamba_radix_cache`, `unified_radix_cache` and `radix_cache_cpp` all slice the kv row by `kv_len_to_handle` directly. `RadixCache` was the only backend deriving the range from token IDs.
- That made the release range depend on `kv_committed_len <= len(origin_input_ids) + len(output_ids)`. Under overlap scheduling `kv_committed_len` is advanced in `prepare_for_decode` before the forward runs, so the two can differ transiently. No cleanup entry point is known to be reachable inside that window today, so this is defensive rather than a fix for an observed leak.

## Ownership contract

- `[0, cache_protected_len)` is cache-owned and must be preserved.
- `[cache_protected_len, kv_len_to_handle)` is request-owned committed KV and is released on no-insert cleanup.
- Anything past `kv_len_to_handle` is released by `release_kv_cache`.

## Change

- Return early for `is_insert=False`, releasing `[cache_protected_len, kv_len_to_handle)` with the existing `free_segment` primitive and dropping the radix-node lock, mirroring the `disable` branch directly above it.
- The insertion path is unchanged and no longer needs the `if is_insert:` guard.
- Extend the matched-prefix failure test to cover a committed KV slot with no corresponding output token.

## Validation

- `test_decode_radix_lock_ref.py`, `test_radix_cache_unit.py`, `test_paged_free_segment.py`, `test_free_kv_row_coalesce.py`, `test_unified_radix_lock_ref.py`: 69 tests, all passing.
- Page-set equivalence checked directly against `PagedTokenToKVPoolAllocator` with `page_size=4`: across every `(sequence length, cache_protected_len)` pair, the single-segment release frees exactly the same pages as the previous two-segment `free_segments` call, and the follow-up overallocated-tail release never double-frees a page.
- Repository pre-commit checks pass for both changed files.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #35288165906](https://github.com/sgl-project/sglang/actions/runs/35288165906)
Latest PR Test (Extra): :x: [Run #35288165706](https://github.com/sgl-project/sglang/actions/runs/35288165706)
Latest PR Test (AMD ROCm 10): :x: [Run #35288165863](https://github.com/sgl-project/sglang/actions/runs/35288165863)

### CODE REVIEW

**Comment by gongy** [comment]
/rerun-test test/registered/unit/mem_cache/test_decode_radix_lock_ref.py

**Review by guapisolo — APPROVED** [review]
good fix. Need code owner review.

**Comment by nanjiangwill** [comment]
Repro against current `main` after the scheduler refactor

We found this while auditing #34037 cleanup: overlap scheduling can commit a KV slot before the sampled token is accepted. The existing incremental-transfer-failure test reaches `is_insert=False`, but keeps token count equal to committed-KV length and checks lock/prefix state rather than the allocator release range.

This CPU-only script constructs the missing state directly:

```python
import runpy
from array import array
from pathlib import Path

import torch

from sglang.srt.mem_cache.base_prefix_cache import MatchPrefixParams
from sglang.srt.mem_cache.radix_cache import RadixKey

test_module = runpy.run_path(
    Path.cwd()
    / "test/registered/unit/mem_cache/test_decode_radix_lock_ref.py"
)
TestDecodeLockRefScenarios = test_module["TestDecodeLockRefScenarios"]
_make_cache_with_pools = test_module["_make_cache_with_pools"]
_make_req = test_module["_make_req"]

cache, req_to_token = _make_cache_with_pools()

prefix_ids = [1, 2, 3]
prefix_kv = [10, 20, 30]
TestDecodeLockRefScenarios()._populate_prefix(cache, prefix_ids, prefix_kv)

match = cache.match_prefix(
    MatchPrefixParams(key=RadixKey(array("q", prefix_ids)))
)
prefix_len = len(match.device_indices)
cache.inc_lock_ref(match.last_device_node)

# The prefix owns 10, 20, 30. The request owns 40 and 50.
all_token_ids = [1, 2, 3, 4, 5]
all_kv = [10, 20, 30, 40, 50]
req_to_token[0, : len(all_kv)] = torch.tensor(all_kv, dtype=torch.int64)

req = _make_req(
    fill_ids=all_token_ids,
    req_pool_idx=0,
    cache_protected_len=prefix_len,
    last_node=match.last_device_node,
)

# Legal overlap state: KV for token 5 is committed, but token 5 has not yet
# been accepted into output_ids.
req.output_ids = array("q")
kv_committed_len = getattr(req.kv, "kv_committed_len", None)
if kv_committed_len is None:
    kv_committed_len = req.kv_committed_len

assert kv_committed_len == 5
assert len(req.origin_input_ids + req.output_ids) == 4

allocator = cache.token_to_kv_pool_allocator
allocator.reset_mock()

cache.cache_finished_req(
    req,
    is_insert=False,
    kv_len_to_handle=kv_committed_len,
)

freed = []
for call in allocator.free_segment.call_args_list:
    freed.extend(call.args[0].tolist())
for call in allocator.free_segments.call_args_list:
    for indices, _start_pos in call.args[0]:
        freed.extend(indices.tolist())

assert freed == [40, 50], (
    f"expected all request-owned KV [40, 50] to be released, got {freed}; "
    "the final overlap-committed slot leaked"
)

print("PASS: no-insert cleanup released both request-owned KV slots")
```

Run from the repository root:

```bash
PYTHONPATH=python:. python repro_35204.py
```

Verified results:

- Current `main` at `d249672ad3`: fails with `got [40]`; slot `50` is not released.
- This PR branch: passes and releases `[40, 50]`.

The scheduler refactor moved the ownership fields but did not change this physical-versus-logical boundary: `kv_len_to_handle` is 5 while the accepted token sequence is still length 4.

**Comment by hnyls2002** [comment]
/rerun-test test_decode_radix_lock_ref.py test_radix_cache_unit.py test_paged_free_segment.py test_free_kv_row_coalesce.py test_unified_radix_lock_ref.py
